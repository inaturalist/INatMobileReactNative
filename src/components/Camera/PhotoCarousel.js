// @flow

import { useNavigation } from "@react-navigation/native";
import classnames from "classnames";
import { ImageBackground, Pressable, View } from "components/styledComponents";
import type { Node } from "react";
import React, { useEffect, useState } from "react";
import {
  ActivityIndicator,
  FlatList
} from "react-native";
import LinearGradient from "react-native-linear-gradient";
import Modal from "react-native-modal";
import { IconButton, useTheme } from "react-native-paper";

type Props = {
  emptyComponent?: Function,
  selectedPhotoIndex?: number,
  containerStyle?: string,
  savingPhoto?: boolean,
  deletePhoto?: Function,
  isLandscapeMode?:boolean,
  isLargeScreen?: boolean,
  setMediaViewerUris: Function,
  photoUris: Array<string>,
  setSelectedPhotoIndex: Function
}

const PhotoCarousel = ( {
  emptyComponent,
  selectedPhotoIndex,
  containerStyle,
  savingPhoto,
  deletePhoto,
  isLandscapeMode,
  isLargeScreen,
  setMediaViewerUris,
  photoUris,
  setSelectedPhotoIndex
}: Props ): Node => {
  const theme = useTheme( );
  const navigation = useNavigation( );
  const [deletePhotoMode, setDeletePhotoMode] = useState( false );
  const imageClass = "justify-center items-center";
  const smallPhotoClass = "rounded-sm w-[42px] h-[42px]";
  const largePhotoClass = "rounded-md w-[83px] h-[83px]";

  useEffect( () => {
    if ( photoUris.length === 0 && deletePhotoMode ) {
      setDeletePhotoMode( false );
    }
  }, [photoUris.length, deletePhotoMode] );

  const renderSkeleton = ( ) => ( savingPhoto
    ? (
      <View
        className={classnames(
          "flex",
          {
            "w-fit h-full": isLargeScreen && isLandscapeMode
          },
          imageClass
        )}
      >
        <View
          className={classnames(
            "bg-lightGray justify-center",
            {
              [`${smallPhotoClass} mx-[3px]`]: !isLargeScreen,
              [`${largePhotoClass} mx-[8.5px]`]: isLargeScreen && !isLandscapeMode,
              [`${largePhotoClass}`]: isLargeScreen && isLandscapeMode
            }
          )}
        >
          <ActivityIndicator />
        </View>
      </View>
    )
    : null );

  const renderPhotoOrEvidenceButton = ( { item, index } ) => (
    <>
      <Pressable
        accessibilityRole="button"
        onLongPress={( ) => {
          if ( deletePhoto ) {
            setDeletePhotoMode( mode => !mode );
          }
        }}
        onPress={( ) => {
          if ( deletePhotoMode && deletePhoto ) {
            deletePhoto( item );
          } else {
            setSelectedPhotoIndex( index );
            setMediaViewerUris( photoUris );
            navigation.navigate( "MediaViewer" );
          }
        }}
        className={classnames(
          imageClass,
          {
            "mt-12": containerStyle === "camera",
            "mt-6": containerStyle !== "camera",
            "border border-selectionGreen border-4": selectedPhotoIndex === index
          },
          {
            "mx-[3px] mt-0": !isLargeScreen,
            "mx-[8.5px] mt-0": isLargeScreen && isLandscapeMode,
            "my-[18px] mt-0": isLargeScreen && !isLandscapeMode
          }
        )}
      >
        <View
          testID="PhotoCarousel.photo"
          className={classnames(
            "overflow-hidden",
            {
              [`${smallPhotoClass}`]: !isLargeScreen,
              [`${largePhotoClass}`]: isLargeScreen
            }
          )}
        >
          <ImageBackground
            source={{ uri: item }}
            className={classnames(
              `w-fit h-full flex ${imageClass}`
            )}
          >
            {deletePhotoMode && (
              <LinearGradient
                className="absolute inset-0"
                colors={["rgba(0, 0, 0, 0.5)", "rgba(0, 0, 0, 0.5)"]}
              />
            )}
            {( containerStyle === "camera" && deletePhotoMode ) && (
              <IconButton
                icon="trash-outline"
                mode="contained-tonal"
                iconColor={theme.colors.onPrimary}
                containerColor="rgba(0, 0, 0, 0.5)"
                size={30}
              />
            )}
          </ImageBackground>
        </View>
      </Pressable>
      {index === photoUris.length - 1 && renderSkeleton( )}
    </>
  );

  const photoPreviewsList = (
    <FlatList
      data={[...photoUris]}
      renderItem={renderPhotoOrEvidenceButton}
      horizontal={!isLargeScreen || !!isLandscapeMode}
      ListEmptyComponent={savingPhoto
        ? renderSkeleton( )
        : emptyComponent}
    />
  );

  return deletePhotoMode
    ? (
      <Modal
        visible
        onBackdropPress={() => setDeletePhotoMode( false )}
        backdropOpacity={0}
        // eslint-disable-next-line react-native/no-inline-styles
        style={{ margin: 0 }}
      >
        <View className={classnames(
          "absolute top-0 pt-[50px]",
          {
            "ml-[18px]": isLargeScreen && isLandscapeMode
          }
        )}
        >
          {photoPreviewsList}
        </View>
      </Modal>
    )
    : photoPreviewsList;
};

export default PhotoCarousel;
