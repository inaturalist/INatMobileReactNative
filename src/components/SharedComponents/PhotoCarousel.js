// @flow

import type { Node } from "react";
import React from "react";
import {
  ActivityIndicator, FlatList, Image, Pressable, View
} from "react-native";
import { Avatar, useTheme } from "react-native-paper";
import { useSafeAreaInsets } from "react-native-safe-area-context";

import { imageStyles, viewStyles } from "../../styles/sharedComponents/photoCarousel";

type Props = {
  emptyComponent?: Function,
  photoUris: Array<string>,
  setSelectedPhotoIndex?: Function,
  selectedPhotoIndex?: number,
  containerStyle?: string,
  handleDelete?: Function,
  savingPhoto?: boolean
}

const PhotoCarousel = ( {
  photoUris,
  emptyComponent,
  setSelectedPhotoIndex,
  selectedPhotoIndex,
  containerStyle,
  handleDelete,
  savingPhoto
}: Props ): Node => {
  const insets = useSafeAreaInsets( );
  const { colors } = useTheme( );
  const renderDeleteButton = photoUri => (
    <Pressable
      onPress={( ) => {
        if ( !handleDelete ) { return; }
        handleDelete( photoUri );
      }}
      style={viewStyles.deleteButton}
    >
      <Avatar.Icon icon="delete-forever" size={30} style={{ backgroundColor: colors.background }} />
    </Pressable>
  );

  const renderSkeleton = ( ) => {
    if ( savingPhoto ) {
      return (
        <View style={viewStyles.photoLoading}>
          <ActivityIndicator />
        </View>
      );
    }
    return null;
  };

  const renderPhoto = ( { item, index } ) => (
    <>
      <Pressable
        onPress={( ) => {
          if ( setSelectedPhotoIndex ) {
            setSelectedPhotoIndex( index );
          }
        }}
      >
        <Image
          source={{ uri: item }}
          style={[
            imageStyles.photo,
            selectedPhotoIndex === index && viewStyles.greenSelectionBorder,
            ( containerStyle === "camera" ) && imageStyles.photoStandardCamera
          ]}
          testID="ObsEdit.photo"
        />
        {( containerStyle === "camera" ) && renderDeleteButton( item )}
      </Pressable>
      {index === photoUris.length - 1 && renderSkeleton( )}
    </>
  );

  return (
    <FlatList
      data={photoUris}
      contentContainerStyle={( containerStyle === "camera" ) && [
        viewStyles.photoContainer, {
          top: insets.top
        }]}
      renderItem={renderPhoto}
      horizontal
      ListEmptyComponent={savingPhoto ? renderSkeleton( ) : emptyComponent}
    />
  );
};

export default PhotoCarousel;
