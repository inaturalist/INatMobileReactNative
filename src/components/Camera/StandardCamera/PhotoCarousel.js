// @flow

import classnames from "classnames";
import MediaViewerModal from "components/MediaViewer/MediaViewerModal";
import { INatIconButton } from "components/SharedComponents";
import { ImageBackground, Pressable, View } from "components/styledComponents";
import { RealmContext } from "providers/contexts";
import type { Node } from "react";
import React, {
  useCallback, useEffect, useRef, useState
} from "react";
import {
  ActivityIndicator,
  FlatList
} from "react-native";
import Modal from "react-native-modal";
import { useTheme } from "react-native-paper";
import Animated, {
  useAnimatedStyle,
  withTiming
} from "react-native-reanimated";
import ObservationPhoto from "realmModels/ObservationPhoto";
import { useTranslation } from "sharedHooks";
import useStore from "stores/useStore";

const { useRealm } = RealmContext;

type Props = {
  emptyComponent?: Function,
  takingPhoto?: boolean,
  isLandscapeMode?:boolean,
  isLargeScreen?: boolean,
  isTablet?: boolean,
  rotation?: {
    value: number
  },
  photoUris: Array<string>
}

export const SMALL_PHOTO_DIM = 42;
export const LARGE_PHOTO_DIM = 83;
export const SMALL_PHOTO_GUTTER = 6;
export const LARGE_PHOTO_GUTTER = 17;
const IMAGE_CONTAINER_CLASSES = ["justify-center", "items-center"];
const SMALL_PHOTO_CLASSES = [
  "rounded-sm",
  "w-[42px]",
  "h-[42x]",
  "mx-[3px]"
];
const LARGE_PHOTO_CLASSES = [
  "rounded-md",
  "w-[83px]",
  "h-[83px]",
  "m-[8.5px]"
];

const PhotoCarousel = ( {
  emptyComponent,
  takingPhoto,
  isLandscapeMode,
  isLargeScreen,
  isTablet,
  rotation,
  photoUris
}: Props ): Node => {
  const deletePhotoFromObservation = useStore( state => state.deletePhotoFromObservation );
  const realm = useRealm( );
  const { t } = useTranslation( );
  const theme = useTheme( );
  const [deletePhotoMode, setDeletePhotoMode] = useState( false );
  const [tappedPhotoIndex, setTappedPhotoIndex] = useState( -1 );
  const photoClasses = isLargeScreen
    ? LARGE_PHOTO_CLASSES
    : SMALL_PHOTO_CLASSES;
  const photoDim = isLargeScreen
    ? LARGE_PHOTO_DIM
    : SMALL_PHOTO_DIM;
  const photoGutter = isLargeScreen
    ? LARGE_PHOTO_GUTTER
    : SMALL_PHOTO_GUTTER;

  useEffect( () => {
    if ( photoUris.length === 0 && deletePhotoMode ) {
      setDeletePhotoMode( false );
    }
  }, [photoUris.length, deletePhotoMode] );

  // I tried passing this in as a prop but the animation wasn't as smooth
  const animatedStyle = useAnimatedStyle(
    () => ( {
      transform: [
        {
          rotateZ: rotation
            ? withTiming( `${-1 * rotation.value}deg` )
            : 0
        }
      ]
    } ),
    [rotation?.value]
  );

  const renderSkeleton = useCallback( ( ) => ( takingPhoto
    ? (
      <View
        className={classnames(
          "flex",
          {
            "w-fit h-full": isTablet && isLandscapeMode
          },
          ...IMAGE_CONTAINER_CLASSES
        )}
      >
        <View
          className={classnames(
            "bg-lightGray justify-center",
            ...photoClasses
          )}
        >
          <ActivityIndicator />
        </View>
      </View>
    )
    : null ), [
    isTablet,
    isLandscapeMode,
    photoClasses,
    takingPhoto
  ] );

  const showDeletePhotoMode = useCallback( ( ) => {
    if ( deletePhotoFromObservation ) {
      setDeletePhotoMode( mode => !mode );
    }
  }, [deletePhotoFromObservation] );

  const viewPhotoAtIndex = useCallback( ( item, index ) => {
    setTappedPhotoIndex( index );
  }, [
    setTappedPhotoIndex
  ] );

  const deletePhotoByUri = useCallback( async photoUri => {
    if ( !deletePhotoFromObservation ) return;
    deletePhotoFromObservation( photoUri );
    await ObservationPhoto.deletePhoto( realm, photoUri );
  }, [deletePhotoFromObservation, realm] );

  const renderPhotoOrEvidenceButton = useCallback( ( { item: photoUri, index } ) => (
    <>
      <Animated.View style={!isTablet && animatedStyle}>
        <View
          className={classnames( IMAGE_CONTAINER_CLASSES )}
        >
          <View
            testID="PhotoCarousel.photo"
            className={classnames(
              "overflow-hidden",
              ...photoClasses
            )}
          >
            <ImageBackground
              source={{ uri: photoUri }}
              className={classnames(
                "w-fit",
                "h-full",
                "flex",
                ...IMAGE_CONTAINER_CLASSES
              )}
            >
              {
                deletePhotoMode
                  ? (
                    <View
                      className="w-full h-full flex-1 justify-center items-center bg-black/50"
                    >
                      <INatIconButton
                        icon="trash-outline"
                        mode="contained"
                        color={theme.colors.onPrimary}
                        backgroundColor="rgba(0, 0, 0, 0.5)"
                        testID={`PhotoCarousel.deletePhoto.${photoUri}`}
                        accessibilityLabel={t( "Delete-photo" )}
                        onPress={( ) => deletePhotoByUri( photoUri )}
                      />
                    </View>
                  )
                  : (
                    <Pressable
                      accessibilityRole="button"
                      accessibilityLabel={t( "View-photo" )}
                      testID={`PhotoCarousel.displayPhoto.${photoUri}`}
                      onLongPress={showDeletePhotoMode}
                      onPress={( ) => viewPhotoAtIndex( photoUri, index )}
                      className="w-full h-full"
                    />
                  )
              }
            </ImageBackground>
          </View>
        </View>
      </Animated.View>
      {index === photoUris.length - 1 && renderSkeleton( )}
    </>
  ), [
    animatedStyle,
    deletePhotoByUri,
    deletePhotoMode,
    isTablet,
    photoClasses,
    photoUris,
    renderSkeleton,
    showDeletePhotoMode,
    t,
    theme,
    viewPhotoAtIndex
  ] );

  const photoPreviewsList = (
    <FlatList
      data={[...photoUris]}
      renderItem={renderPhotoOrEvidenceButton}
      horizontal={!isTablet || !isLandscapeMode}
      ListEmptyComponent={takingPhoto
        ? renderSkeleton( )
        : emptyComponent}
    />
  );

  // Mild acrobatics to position the photos in delete mode inside a modal
  // exactly the same place they appear normally. The modal is useful to,
  // well, behave like a modal, because in delete mode you can only delete or
  // exit delete moda, and taps outside the photos exit delete mode. But the
  // modal exists outside the normal view hierarchy so you can't position
  // things relative to containing components. Instead, we're using a ref and
  // the measure() method to store the position of the container element in
  // state, and use that to position another container inside the modal in
  // exactly the same place

  const containerRef = useRef( );
  const [containerPos, setContainerPos] = useState( { x: null, y: null } );
  const containerStyle = {
    height: isTablet && isLandscapeMode
      ? photoUris.length * ( photoDim + photoGutter ) + photoGutter
      : photoDim + ( photoGutter * 2 ),
    padding: photoGutter / 2
  };

  return (
    <View
      ref={containerRef}
      onLayout={
        // When the container gets rendered, we store its position on screen
        // in state so we can layout content inside the modal in exactly the
        // same position
        ( ) => containerRef?.current?.measure(
          ( _x, _y, w, h, pageX, pageY ) => setContainerPos( {
            x: pageX,
            y: pageY,
            w,
            h
          } )
        )
      }
      // Dynamic calculation of these values kind of just doesn't work with tailwind.
      // eslint-disable-next-line react-native/no-inline-styles
      style={containerStyle}
    >
      {
        // Render a model version of the previews over the actual previews.
        // Awkward, but it helps handle some weird styling in a landscape
        // tablet
        deletePhotoMode
          ? (
            <Modal
              visible
              onBackdropPress={() => setDeletePhotoMode( false )}
              backdropOpacity={0}
              // We want this to take over the whole screen
              // eslint-disable-next-line react-native/no-inline-styles
              style={{ margin: 0, position: "relative" }}
            >
              <View
                // These layout values need to be dynamic relative to the
                // position of the container outside of the modal
                // eslint-disable-next-line react-native/no-inline-styles
                style={{
                  position: "absolute",
                  left: containerPos.x,
                  top: containerPos.y,
                  ...containerStyle
                }}
              >
                { photoPreviewsList }
              </View>
            </Modal>
          )
          : photoPreviewsList
      }
      <MediaViewerModal
        editable
        showModal={tappedPhotoIndex >= 0}
        onClose={( ) => setTappedPhotoIndex( -1 )}
        onDelete={async photoUri => {
          await deletePhotoByUri( photoUri );
          setTappedPhotoIndex( tappedPhotoIndex - 1 );
        }}
        uri={photoUris[tappedPhotoIndex]}
        photos={photoUris.map( uri => ( { url: uri } ) )}
      />
    </View>
  );
};

export default PhotoCarousel;
