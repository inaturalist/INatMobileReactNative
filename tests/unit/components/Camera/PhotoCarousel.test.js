import { faker } from "@faker-js/faker";
import { render, screen } from "@testing-library/react-native";
import PhotoCarousel from "components/Camera/StandardCamera/PhotoCarousel";
import React from "react";

import * as mockReactI18next from "../../../mocks/react-i18next";

jest.mock( "react-i18next", ( ) => mockReactI18next );

const mockPhotoUris = [
  faker.image.imageUrl( ),
  faker.image.imageUrl( ),
  faker.image.imageUrl( )
];

describe( "PhotoCarousel", ( ) => {
  // There were some tests of photo sizes responding to the isLargeScreen prop
  // but somewhat dynamic tailwind classes don't seem to create style props
  // in a test environment, so I'm not sure we can test those now ~~~kueda
  // 20230518
  it( "renders correctly", async () => {
    render(
      <PhotoCarousel photoUris={mockPhotoUris} />
    );

    // Snapshot test
    expect( screen ).toMatchSnapshot();
  } );
  it( "renders correctly for large screen", async () => {
    render(
      <PhotoCarousel photoUris={mockPhotoUris} isLargeScreen />
    );

    // Snapshot test
    expect( screen ).toMatchSnapshot();
  } );
} );
