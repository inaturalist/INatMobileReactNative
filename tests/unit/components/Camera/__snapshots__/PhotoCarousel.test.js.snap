// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoCarousel renders correctly 1`] = `
<RCTScrollView
  data={
    [
      "https://loremflickr.com/640/480",
      "https://loremflickr.com/640/480",
      "https://loremflickr.com/640/480",
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "justifyContent": "center",
              },
              {
                "alignItems": "center",
              },
              {
                "marginTop": 24,
              },
              {
                "marginLeft": 3,
                "marginRight": 3,
              },
              {
                "marginTop": 0,
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderBottomLeftRadius": 4,
                  "borderBottomRightRadius": 4,
                  "borderTopLeftRadius": 4,
                  "borderTopRightRadius": 4,
                },
                {
                  "width": 42,
                },
                {
                  "height": 42,
                },
              ],
            ]
          }
          testID="PhotoCarousel.photo"
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              [
                [
                  {
                    "height": "100%",
                  },
                  {
                    "display": "flex",
                  },
                  {
                    "justifyContent": "center",
                  },
                  {
                    "alignItems": "center",
                  },
                ],
              ]
            }
          >
            <Image
              source={
                {
                  "uri": "https://loremflickr.com/640/480",
                }
              }
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "height": "100%",
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "justifyContent": "center",
              },
              {
                "alignItems": "center",
              },
              {
                "marginTop": 24,
              },
              {
                "marginLeft": 3,
                "marginRight": 3,
              },
              {
                "marginTop": 0,
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderBottomLeftRadius": 4,
                  "borderBottomRightRadius": 4,
                  "borderTopLeftRadius": 4,
                  "borderTopRightRadius": 4,
                },
                {
                  "width": 42,
                },
                {
                  "height": 42,
                },
              ],
            ]
          }
          testID="PhotoCarousel.photo"
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              [
                [
                  {
                    "height": "100%",
                  },
                  {
                    "display": "flex",
                  },
                  {
                    "justifyContent": "center",
                  },
                  {
                    "alignItems": "center",
                  },
                ],
              ]
            }
          >
            <Image
              source={
                {
                  "uri": "https://loremflickr.com/640/480",
                }
              }
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "height": "100%",
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "justifyContent": "center",
              },
              {
                "alignItems": "center",
              },
              {
                "marginTop": 24,
              },
              {
                "marginLeft": 3,
                "marginRight": 3,
              },
              {
                "marginTop": 0,
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderBottomLeftRadius": 4,
                  "borderBottomRightRadius": 4,
                  "borderTopLeftRadius": 4,
                  "borderTopRightRadius": 4,
                },
                {
                  "width": 42,
                },
                {
                  "height": 42,
                },
              ],
            ]
          }
          testID="PhotoCarousel.photo"
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              [
                [
                  {
                    "height": "100%",
                  },
                  {
                    "display": "flex",
                  },
                  {
                    "justifyContent": "center",
                  },
                  {
                    "alignItems": "center",
                  },
                ],
              ]
            }
          >
            <Image
              source={
                {
                  "uri": "https://loremflickr.com/640/480",
                }
              }
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "height": "100%",
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`PhotoCarousel renders correctly for large screen 1`] = `
<RCTScrollView
  data={
    [
      "https://loremflickr.com/640/480",
      "https://loremflickr.com/640/480",
      "https://loremflickr.com/640/480",
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "justifyContent": "center",
              },
              {
                "alignItems": "center",
              },
              {
                "marginTop": 24,
              },
              {
                "marginBottom": 18,
                "marginTop": 18,
              },
              {
                "marginTop": 0,
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderBottomLeftRadius": 9,
                  "borderBottomRightRadius": 9,
                  "borderTopLeftRadius": 9,
                  "borderTopRightRadius": 9,
                },
                {
                  "width": 83,
                },
                {
                  "height": 83,
                },
              ],
            ]
          }
          testID="PhotoCarousel.photo"
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              [
                [
                  {
                    "height": "100%",
                  },
                  {
                    "display": "flex",
                  },
                  {
                    "justifyContent": "center",
                  },
                  {
                    "alignItems": "center",
                  },
                ],
              ]
            }
          >
            <Image
              source={
                {
                  "uri": "https://loremflickr.com/640/480",
                }
              }
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "height": "100%",
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "justifyContent": "center",
              },
              {
                "alignItems": "center",
              },
              {
                "marginTop": 24,
              },
              {
                "marginBottom": 18,
                "marginTop": 18,
              },
              {
                "marginTop": 0,
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderBottomLeftRadius": 9,
                  "borderBottomRightRadius": 9,
                  "borderTopLeftRadius": 9,
                  "borderTopRightRadius": 9,
                },
                {
                  "width": 83,
                },
                {
                  "height": 83,
                },
              ],
            ]
          }
          testID="PhotoCarousel.photo"
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              [
                [
                  {
                    "height": "100%",
                  },
                  {
                    "display": "flex",
                  },
                  {
                    "justifyContent": "center",
                  },
                  {
                    "alignItems": "center",
                  },
                ],
              ]
            }
          >
            <Image
              source={
                {
                  "uri": "https://loremflickr.com/640/480",
                }
              }
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "height": "100%",
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityRole="button"
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "justifyContent": "center",
              },
              {
                "alignItems": "center",
              },
              {
                "marginTop": 24,
              },
              {
                "marginBottom": 18,
                "marginTop": 18,
              },
              {
                "marginTop": 0,
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderBottomLeftRadius": 9,
                  "borderBottomRightRadius": 9,
                  "borderTopLeftRadius": 9,
                  "borderTopRightRadius": 9,
                },
                {
                  "width": 83,
                },
                {
                  "height": 83,
                },
              ],
            ]
          }
          testID="PhotoCarousel.photo"
        >
          <View
            accessibilityIgnoresInvertColors={true}
            style={
              [
                [
                  {
                    "height": "100%",
                  },
                  {
                    "display": "flex",
                  },
                  {
                    "justifyContent": "center",
                  },
                  {
                    "alignItems": "center",
                  },
                ],
              ]
            }
          >
            <Image
              source={
                {
                  "uri": "https://loremflickr.com/640/480",
                }
              }
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "height": "100%",
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
